Support merging DataFrames on a combo of columns and index levels (GH 14355)

jonmmease opened the pull request:
This PR implements the changes proposed and discussed with @jorisvandenbossche @shoyer @TomAugspurger @jreback in #14355.

These changes allow the `on`, `left_on`, and `right_on` parameters of `DataFrame.merge` to accept a combination of column names and index level names. Any common index levels that are merged on are preserved as index levels in the resulting DataFrame, while all other index levels are removed.

In the case of a conflict, the column takes precedence and an ambiguity `FutureWarning` is raised.

- [x] closes #14355
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] documentation updated
- [x] whatsnew entry

**Note:** The new `df._get_column_or_level_values` method introduced in this PR is the same method introduced in #17361 to support sorting DataFrames on a combination of columns and index levels. I will keep this method in sync between the two PRs during the review process.

Commit:
Support merging frames on a combo of columns and index levels (GH 14355)

gfyoung commented on doc/source/merging.rst:
```diff
@@ -1120,6 +1122,53 @@ This is not Implemented via ``join`` at-the-moment, however it can be done using
+.. note::
+
+   When DataFrames are merged on a string that matches an index level in both
+   frames then the index level is preserved as an index level in the resulting
```
frames --> frames,

gfyoung commented on doc/source/merging.rst:
```diff
@@ -1120,6 +1122,53 @@ This is not Implemented via ``join`` at-the-moment, however it can be done using
+
+.. note::
+
+   If a string matches both a column name and an index level name then a warning is
```
index level name --> index level name,

gfyoung commented on doc/source/whatsnew/v0.21.0.txt:
```diff
@@ -109,6 +109,7 @@ Other Enhancements
 - :func:`date_range` now accepts 'Y' in addition to 'A' as an alias for end of year (:issue:`9313`)
 - Integration with `Apache Parquet <https://parquet.apache.org/>`__, including a new top-level :func:`read_parquet` and :func:`DataFrame.to_parquet` method, see :ref:`here <io.parquet>`. (:issue:`15838`, :issue:`17438`)
 - :func:`DataFrame.add_prefix` and :func:`DataFrame.add_suffix` now accept strings containing the '%' character. (:issue:`17151`)
+- :func:`DataFrame.merge` now accepts index level names as `on`, `left_on`, and `right_on` parameters allowing frames to be merged on a combination of columns and index levels (:issue:`14355`)
```
"parameters allowing" --> "parameters, allowing"

gfyoung commented on pandas/core/frame.py:
```diff
@@ -3437,6 +3437,36 @@ def f(vals):
     # ----------------------------------------------------------------------
     # Sorting
+    def _get_column_or_level_values(self, key, axis=1,
+                                    op_description='retrieve'):
```
Add doc-string for this.  Developers in the future will thank you. 😉

gfyoung commented on pandas/core/frame.py:
```diff
@@ -3437,6 +3437,36 @@ def f(vals):
+                    ("'%s' is both a column name and an index level.\n"
+                     "Defaulting to column but "
+                     "this will raise an ambiguity error in a "
+                     "future version") % key,
```
The percentile string replacement is being phased out in favor of `.format`.

gfyoung commented on pandas/core/reshape/merge.py:
```diff
@@ -792,6 +815,10 @@ def _get_merge_keys(self):
         is_rkey = lambda x: isinstance(
             x, (np.ndarray, Series)) and len(x) == len(right)
 
+        def get_key_vals(df, key):
```
Add doc-string for this.

gfyoung commented on pandas/tests/reshape/test_merge.py:
```diff
@@ -1350,6 +1358,131 @@ def f():
             household.join(log_return, how='outer')
         pytest.raises(NotImplementedError, f)
 
+    def test_merge_on_index_and_column(self):
```
Reference issue number.

gfyoung commented on pandas/tests/reshape/test_merge.py:
```diff
@@ -1350,6 +1358,131 @@ def f():
         pytest.raises(NotImplementedError, f)
 
+    def test_merge_on_index_and_column(self):
+        # Construct DataFrames
```
Reference issue number.

Commit:
Cleanup for review

jmmease commented:
Thanks for the quick feedback @gfyoung. I think I've incorporated all of it in this last commit. Let me know if I've missed anything or if there are additional changes you'd like to see.

jreback commented on doc/source/merging.rst:
```diff
@@ -1120,6 +1122,53 @@ This is not Implemented via ``join`` at-the-moment, however it can be done using
 
+Merging on a combination of columns and index levels
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+.. versionadded:: 0.21
```
need a blank line here

add a ``:ref:`` entry before the sub-section

jmmease replied:
done

jreback commented on pandas/core/frame.py:
```diff
@@ -3437,6 +3437,74 @@ def f(vals):
+
+        Parameters
+        ----------
+        key: int or object
```
This should *only* accept string keys referring to a named column or level, We do not want to have the integer/string mess like ``.ix`` again.

jreback commented on pandas/core/reshape/merge.py:
```diff
@@ -717,7 +719,25 @@ def _maybe_add_join_keys(self, result, left_indexer, right_indexer):
                     result[name] = key_col
                 else:
-                    result.insert(i, name or 'key_{i}'.format(i=i), key_col)
+                    if name and name in result.index.names:
```
you are adding an amazing amount of logic here. This must be  made simpler.

jreback commented on pandas/core/reshape/merge.py:
```diff
@@ -792,6 +815,29 @@ def _get_merge_keys(self):
+            See Also
+            --------
+            DataFrame._get_column_or_level_values
+            """
```
if you actually need complexity, then hide it away here

jreback commented on pandas/core/reshape/merge.py:
```diff
@@ -792,6 +815,29 @@ def _get_merge_keys(self):
+
+            Returns
+            -------
+            values: array
```
array -> np.ndarray

jreback commented on pandas/core/reshape/merge.py:
```diff
@@ -792,6 +815,29 @@ def _get_merge_keys(self):
+            Parameters
+            ----------
+            df: DataFrame
+            key: int or object
```
see above

jreback commented on pandas/core/frame.py:
```diff
@@ -3437,6 +3437,74 @@ def f(vals):
+        axis: int, default 0
+            Axis to retrieve values from
+
+        op_description: str,  default 'retrieve'
```
not sure what this parameter is meant to do, its completely non-intuitive.

jreback reviewed (changes requested):
lots of comments

jreback commented:
We may need a common framework here to incorporate the ``.groupby`` functionaility of the same name. Adding this many code branches is making things pretty complex.

jmmease commented:
Thanks for the feedback @jreback. I'll address your comments and take another stab at driving down the complexity. If you, or anyone else, have some ideas for a more unified way to go about this (mixing columns and index levels in merge, sort_values, groupby, etc) I'd be happy to take a step back and consider a change of course.

Commit:
Simplify and refactor column/level logic in merge

Commit:
revert implementation (but keep documentation and tests)

Commit:
PEP8 cleanup

Commit:
Extract column/level ambiguity warning logic into utility method

Commit:
Add newline and add :ref: entry for new doc section

Commit:
docstring / comment cleanup

jmmease commented:
Ok, after meditating on this section of the code again I think I found a simpler approach.  I created a new set of DataFrame helpers to support this change and I refactored the existing groupby column/index logic (#14432) to use these as well. I haven't revisited my ``sort_values`` updates in #17361 yet, but I should be able to rely on these same helpers there too. 

Does this general approach look reasonable @jreback, @jorisvandenbossche, @gfyoung, and @TomAugspurger?  Thanks!

Commit:
Merge branch 'master' into enh_14355

jmmease commented:
I'm having trouble interpreting the circleci failure above. Is this something my changes caused?

gfyoung commented:
@jmmease : The Circle CI builds just died...let's give it another run to see if it passes.

jmmease commented:
Thanks, @gfyoung

gfyoung commented:
Ping @jreback for review now.

jreback commented on pandas/tests/reshape/test_merge.py:
```diff
@@ -1350,6 +1359,138 @@ def f():
             household.join(log_return, how='outer')
         pytest.raises(NotImplementedError, f)
 
```
can you make this a separe test class. Try to use fixtures for these df's as well (and/or parametrize)

jmmease replied:
Done. See ``TestMergeColumnAndIndex``

jreback commented on pandas/core/groupby.py:
```diff
@@ -2776,16 +2776,11 @@ def is_in_obj(gpr):
                 in_axis, name, gpr = True, gpr, obj[gpr]
                 exclusions.append(name)
-            elif gpr in obj.index.names:
+            elif obj._is_index_reference(gpr):
```
this should fail on a Series the way it works currently

jmmease replied:
Thanks for pointing this out @jreback.  I'm still working on a full set of updates for this PR but in the meantime, I added a test case to #17843 that would have caught this.

jreback commented on pandas/core/frame.py:
```diff
@@ -2160,6 +2160,150 @@ def _getitem_frame(self, key):
+            return False
+
+    def _is_index_reference(self, key):
+        """
```
move this to generic.py it should take an axis argument and only look at that axis.

jreback replied:
pls add some specific tests for this

jmmease replied:
I moved all of the new helper methods to generic.py and added axis args.
Tests for all of the new methods added in ``tests/indexes/test_label_or_level_utils.py``

jmmease commented:
Thanks for the comments @jreback, I'll get to them soon.

Commit:
Documentation updates
 - Added whatsnew 0.21 subsection
 - Updated DataFrame.merge() API docstring

Commit:
Fix errors in _drop_columns_or_levels

Commit:
Refactor and parametrize test cases

Commit:
Moved label/level helpers up to NDFrame, added axis support, and added test suite

Commit:
PEP8

jmmease commented on doc/source/whatsnew/v0.21.0.txt:
```diff
@@ -187,6 +187,8 @@ Other Enhancements
 - Integration with `Apache Parquet <https://parquet.apache.org/>`__, including a new top-level :func:`read_parquet` and :func:`DataFrame.to_parquet` method, see :ref:`here <io.parquet>`. (:issue:`15838`, :issue:`17438`)
 - :func:`DataFrame.add_prefix` and :func:`DataFrame.add_suffix` now accept strings containing the '%' character. (:issue:`17151`)
 - Read/write methods that infer compression (:func:`read_csv`, :func:`read_table`, :func:`read_pickle`, and :meth:`~DataFrame.to_pickle`) can now infer from non-string paths, such as ``pathlib.Path`` objects (:issue:`17206`).
+- :func:`DataFrame.merge` now accepts index level names as `on`, `left_on`, and `right_on` parameters, allowing frames to be merged on a combination of columns and index levels (:issue:`14355`)
```
I added a new subsection. Should I remove this one?

jmmease commented:
Checks green. Ready for review @jreback @gfyoung

gfyoung commented on doc/source/merging.rst:
```diff
@@ -504,7 +504,7 @@ the data in DataFrame.
 
 Users who are familiar with SQL but new to pandas might be interested in a
-:ref:`comparison with SQL<compare_with_sql.join>`.
+:ref:`comparison with SQL<F>`.
```
What does "F" refer to?

jmmease replied:
Oh, nothing. Looks like I added that error back when I was adding the `:ref:` entry for my new section. It will be reverted in my next push. Thanks!

Commit:
Revert accidental change to merging.rst

Commit:
Use fixtures for new TestMergeColumnAndIndex tests

Commit:
Merge branch 'master' into enh_14355

# Conflicts:
#	doc/source/whatsnew/v0.21.0.txt

Commit:
Merge branch 'master' into enh_14355

# Conflicts:
#	doc/source/whatsnew/v0.21.0.txt

jmmease commented:
Resolved whatsnew conflict.  At this point should we wait until post-0.21 for this? I can make whatever changes are needed tonight and this weekend if we can still get it in 0.21. If not I'll just remerge and ping after the 0.21 release (Will that be 0.22 or 1.0?). Thanks!
@jreback @TomAugspurger @jorisvandenbossche

jreback commented:
yes, this will wait for post 0.21.0. going to be 0.22 (but no doc-notes yet), so hold off

jmmease commented:
Sounds good, thanks

Commit:
Merge branch 'master' into enh_14355

# Conflicts:
#	doc/source/whatsnew/v0.21.0.txt

Commit:
Update documentation for a 0.22 release

jmmease commented:
Moved whatsnew entry to 0.22. This is ready for continued review as you all have time. Thanks.
@jreback @gfyoung @jorisvandenbossche

gfyoung commented:
@jmmease : Can you rebase onto `master` to get all builds into the green?

Commit:
Merge remote-tracking branch 'upstream/master' into enh_14355

jmmease commented:
I just merged master.  I merged instead of rebased for review continuity, but let me know if you'd prefer a rebase. Thanks.

jreback commented:
> I just merged master. I merged instead of rebased for review continuity, but let me know if you'd prefer a rebase. Thanks.

that's fine, in reality since we rebase & squash on merge anyhow it doesn't make much difference.

jreback commented on doc/source/merging.rst:
```diff
@@ -518,14 +518,16 @@ standard database join operations between DataFrame objects:
-- ``on``: Columns (names) to join on. Must be found in both the left and
-  right DataFrame objects. If not passed and ``left_index`` and
+- ``on``: Column or index level names to join on. Must be found in both the left
+  and right DataFrame objects. If not passed and ``left_index`` and
```
an you add a comment (here and left_on/right_on) that index level merging is new in 0.22.0 (or maybe in a Note section below)

jreback commented on doc/source/whatsnew/v0.22.0.txt:
```diff
@@ -17,6 +17,36 @@ New features
+parameters may now refer to either column names or index level names.  This enables
+merging ``DataFrame`` instances on a combination of index levels and columns
+without resetting indexes. See the :ref:`Merge on columns and levels
+<merging.merge_on_columns_and_levels>` documentation section.
```
add the issue number

jreback commented on pandas/core/frame.py:
```diff
@@ -147,16 +147,17 @@
+    Column or index level names to join on in the right DataFrame. Can also
+    be a vector or list of vectors of the length of the right DataFrame.
+    These vectors are treated as though they are columns.
 left_index : boolean, default False
```
same as above, maybe in a Note section

jreback commented on pandas/core/generic.py:
```diff
@@ -1038,6 +1038,181 @@ def equals(self, other):
+            Axis that levels are associated with (0 for index, 1 for columns)
+
+        Returns
+        -------
```
same as above

jreback commented on pandas/core/generic.py:
```diff
@@ -1038,6 +1038,181 @@ def equals(self, other):
+    # have consistent precedence and validation logic throughout the library.
+
+    _shared_docs['_is_level_reference'] = """
+        Test whether a key is a level reference for a given axis.
```
so this is getting very complicated with implementations both in Series & DataFrame. you actually don't need any of that and instead you can write these methods generically here. you *already* know how many axes you have, and what the info axes are (IOW the primary selection axis). Its going to be way less code (not to mention you don't need to add the shared docs).

jreback replied:
also need to guard against anything with ndim > 2 (just raise).

jreback commented on pandas/tests/indexes/test_label_or_level_utils.py:
```diff
@@ -0,0 +1,378 @@
+    """
+
+    # Setup
+    # =====
```
I would prefer you use fixtures here and not the old style steup_method. (e.g. this is what prepare_df basically is already)

jreback replied:
move this to pandas/tests/generic (new dir I am creating now)

jreback commented on pandas/tests/indexes/test_label_or_level_utils.py:
```diff
@@ -0,0 +1,378 @@
+    def prepare_df1(self, levels=None, axis=0):
+        """Return DataFrame with specified levels (list of any of 'L1', 'L2',
+        and 'L3'). Remaining keys are left as labels"""
+        return self.prepare_df(self.df1, levels=levels, axis=axis)
```
all of these become fixtures

jreback commented on pandas/tests/indexes/test_label_or_level_utils.py:
```diff
@@ -0,0 +1,378 @@
+    @pytest.mark.parametrize('axis', [0, 1])
+    def test_is_level_or_label_reference_df_simple(self, axis):
+
+        # df has no named levels on axis
```
then you just pass them in ax args

jreback commented on pandas/tests/reshape/test_merge.py:
```diff
@@ -69,6 +69,15 @@ def test_merge_common(self):
+        result = merge(left, right, on='key1')
+        expected = merge(self.df, self.df2, on='key1').set_index('key1')
+        assert_frame_equal(result, expected)
+
```
let's add these tests to a new file

Commit:
Documentation updates

Commit:
Moved test_label_or_level_utils to pandas/tests/generic

Commit:
Refactored level_or_level test cases to use fixtures

Commit:
Moved label_or_level utils on Series and DataFrame to NDFrame

Commit:
fix test comment typo

Commit:
PEP8ify

Commit:
Merge remote-tracking branch 'upstream/master' into enh_14355

Commit:
Moved column and index tests to new file

Commit:
Remove test class and convert to using fixtures

Commit:
Rename new test file

jmmease commented:
Thanks, @jreback for the helpful feedback.  I believed I have addressed all of your comments, but let me know if I've missed anything or if there are other changes you'd like to see.

jreback commented on doc/source/whatsnew/v0.22.0.txt:
```diff
@@ -17,6 +17,36 @@ New features
+parameters may now refer to either column names or index level names (:issue:`14355`).
+This enables merging ``DataFrame`` instances on a combination of index levels
+and columns without resetting indexes. See the :ref:`Merge on columns and
+levels <merging.merge_on_columns_and_levels>` documentation section.
```
issue number here

jmmease replied:
issue reference moved to the end of paragraph

jreback commented on pandas/core/generic.py:
```diff
@@ -1038,6 +1038,308 @@ def equals(self, other):
+        axis = self._get_axis_number(axis)
+        other_axes = [ax for ax in range(self._AXIS_LEN) if ax != axis]
+
+        if self.ndim > 2:
```
haven't looked at tests yet, make sure this is tested (the ndim > 2)

jmmease replied:
I added exception assertion tests for all of the new methods against a Panel object.  Does that cover ndim > 2 well enough?

jreback commented on pandas/tests/generic/test_label_or_level_utils.py:
```diff
@@ -0,0 +1,368 @@
+    return expected_labels, expected_levels
+
+
+def check_label_reference(frame, labels, axis):
```
preface these by assert_ in the function name (check is ok, but assert_ is more inline with the codebase)

jmmease replied:
renamed

jreback reviewed (changes requested):
your tests aren't actually being run on the CI because I haven't merged the generic change. soon.

Commit:
Documentation and testing review updates

jreback commented:
i merged the change which adds that generic test dir
so rebase when u can

Commit:
Merge remote-tracking branch 'upstream/master' into enh_14355

jmmease commented:
merged and pushed.

jmmease commented:
Just wanted to follow up on this. Do you think we'll be able to get this in for 0.22?
Thanks for all of the feedback already provided. @jreback @gfyoung @jorisvandenbossche

jreback commented:
@jmmease lgtm. sorry, been flooded. pinging is good way to get attention!

@jorisvandenbossche @TomAugspurger 

I think we could put all of the added code that you put in generic.py in *another* module. maybe references and make it into a mix-in class (that you just include into generic).

?

jreback commented:
> I think we could put all of the added code that you put in generic.py in another module. maybe references and make it into a mix-in class (that you just include into generic).

actually this is ok for now. will think splitting generic.py up (in another issue).

jmmease commented:
No problem @jreback , thanks for all you're doing to keep pandas going strong!

Just let me know if you would like something moved before merging.

Commit:
Merge remote-tracking branch 'upstream/master' into enh_14355

Commit:
Merge remote-tracking branch 'upstream/master' into enh_14355

# Conflicts:
#	doc/source/whatsnew/v0.22.0.txt

jmmease commented:
Just merged to fix whatsnew conflict.  Is this all set @jreback?

jreback commented:
@jmmease small linting error (we just added this, it now looks for list-comprehensions that are not generators) :>

and want @jorisvandenbossche @TomAugspurger to have a look.

Commit:
Fix generator/list lint issues

jmmease commented:
Ok, I think I fixed the lint issues. Thanks for pointing me in the right direction @jreback 

I'll stand-by for any feedback from @jorisvandenbossche and @TomAugspurger

jorisvandenbossche commented on pandas/core/frame.py:
```diff
@@ -147,16 +147,17 @@
-    vectors of the length of the DataFrame to use a particular vector as
-    the join key instead of columns
+    Column or index level names to join on in the left DataFrame. Can also
+    be a vector or list of vectors of the length of the left DataFrame.
```
now you are changing this anyway, can you change vector with array ?

jorisvandenbossche commented on pandas/core/frame.py:
```diff
@@ -147,16 +147,17 @@
-    the join key instead of columns
+    Column or index level names to join on in the left DataFrame. Can also
+    be a vector or list of vectors of the length of the left DataFrame.
+    These vectors are treated as though they are columns.
```
I think 'as if' is easier language than 'as though' for non-native speakers

jorisvandenbossche commented on pandas/core/generic.py:
```diff
@@ -1038,6 +1038,313 @@ def equals(self, other):
+                          level_article=level_article,
+                          level_type=level_type,
+                          label_article=label_article,
+                          label_type=label_type), FutureWarning)
```
this warning needs a stacklevel 

I would also put the creation of the message on a separate line before, I think that will make it cleaner:

```
msg = (....).format(..)
warnings.warn(msg, FutureWarning, stacklevel=..)
```

jorisvandenbossche commented on pandas/tests/reshape/test_merge_index_as_string.py:
```diff
@@ -0,0 +1,193 @@
+                         [['outer'], ['inner'],
+                          ['outer', 'inner'],
+                          ['inner', 'outer']])
+def test_merge_indexes_and_columns_on(left_df, right_df, on, how):
```
This single test adds actually 144 (3x3x4x4) tests, which feels a bit as overkill to me?

jmmease replied:
I trimmed it down by a factor of 4 by removing the `how` fixture and parameterizing `how` alongside `on`.

jorisvandenbossche reviewed (commented):
Nice work!

I added some minor comments.

General notes:

* For the case that an index level and column level match, shouldn't we treat this the same as a duplicate column name (in the future)? (in the idea of seeing an index as a special column) 
  I don't think it changes anything right now (warning that for now column is used), but I would change it in the future to the same as what happens with duplicate columns.

* You wrote the code dealing with selecting labels vs levels very generic. But do we actually have a use case for the `axis=1` ? (selecting labels from index or levels from columns) I think merge is only column-wise? 
  If we don't have such a use case, I think the code would be simpler to just assume it is only for the column labels.

* Should `DataFrame.join` support the same?

jmmease commented:
Thanks for looking things over @jorisvandenbossche. I'll address the specific comments in an update soon but wanted to follow up on a few of your general points.

1) Yes, I agree that it makes sense for ambiguous index/column combinations to eventually fail in the same way that duplicate columns now fail.  Let me know if there's any language in the `FutureWarning` or in my comments in the code that you'd like me to change to make this future intention clearer.

2) I'm planning to use the `axis=1` case for a couple of these new methods in my next update to #17361 (The `sort_values` analog of this PR) since `sort_values` supports that `axis` parameter.  My hope was that these methods would make it easier to add mixed column/index support to other operations in the future, though `sort_values` is the only remaining operation on my personal wishlist at the moment.

3) I restricted index level references to strings at the moment partially because I wasn't clear on how to check for the full set of valid label/level reference types.  I'm happy to open this up if there's a clear way to check this.

4) I'll need to think some more about `join`. Does anyone else have thoughts on this?

Commit:
Allow non-None hashable objects to reference index levels (not just strings)

Commit:
Reduce parameterized test case count by removing how fixture

Now we parameterize the how parameter alongside on. Reduces test-case count by a factor of 4.

Commit:
Refactor warning code and add stacklevel

Commit:
Use single backticks to reference method params in docstrings

Rename vector -> array

Commit:
Add tests and docstring updates for using index levels as `on` param to df.join

Fixed errors that cropped up when using join on a combination of columns and index levels

Commit:
PEP8

jmmease commented on pandas/core/generic.py:
```diff
@@ -1038,6 +1038,313 @@ def equals(self, other):
+                "_is_label_reference is not implemented for {type}"
+                .format(type=type(self)))
+
+        return (isinstance(key, compat.string_types) and
```
Good point, I've changed this to check that the key is non-None and hashable

jmmease commented:
@jorisvandenbossche I just pushed a new version that I believe addresses all of your inline comments (I've responded to a couple of these comments above).

After looking things over I realized that `join` was almost working already. I added some tests, fixed up a few issues, and updated the `join` docstring.

Let me know if there's anything else you'd like to see once you've had a chance to look things over again. Thanks!

jmmease commented:
Just following up. Did this last round of changes address your comments sufficiently @jorisvandenbossche?

@TomAugspurger is there anything more you'd like to see here?

cc: @jreback 

Thanks!

jreback commented on pandas/core/frame.py:
```diff
@@ -195,6 +196,11 @@
 
     .. versionadded:: 0.21.0
 
+Note
```
needs update here

jmmease replied:
Thanks for catching this @jreback. Fixed

jreback reviewed (changes requested):
tiny doc update

Commit:
Fixed Note->Notes in docstring

jmmease commented:
doc change pushed and all green @jreback @jorisvandenbossche @TomAugspurger

jreback commented:
thanks @jmmease 

if any fallout we can PR later!

jmmease commented:
Awesome! thanks @jreback 
I'll be around and be working on #14353 next. Just ping me as any integration issues with this arise
